feat[FieldData]: export to ZBF

bzhangflex opened the pull request:
- Adds a `to_zbf()` method to `FieldData` for exporting E fields to a ZBF file.
- Adds a `from_zbf()` method to `FieldDataset` for importing E fields from a ZBF file.
- Adds a `ZBFData` class for reading ZBF files

Closes https://github.com/flexcompute/tidy3d/issues/2321

bzhangflex committed:
feat[FieldData]: export to and import from ZBF

yaugenst-flex commented on tests/test_data/test_monitor_data.py:
Nice! Two comments on the tests:

1. It looks like you already implemented all the logic for reading a zbf file here. Should we also add a `FieldDataset.from_zbf()`? That'd make it really easy to create custom field sources from zbf data.
2. Instead of re-implementing dummy field data generation here it's probably cleaner to set up a simulation and use the emulated run to get the field data? That'd also be a bit more representative of the actual workflow I guess.

bzhangflex replied:
For 2. are you referring to this for the emulated run? https://github.com/flexcompute/tidy3d/blob/b848f5da93b82bb765fbb86e165db52c89650042/tests/utils.py#L920

yaugenst-flex replied:
yea exactly

yaugenst-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+        fname: str,
+        background_refractive_index: float,
+        freq: Optional[float] = None,
+        mode_index: Optional[int] = 0,
```
Looks like this is not really `Optional` but would always expect an `int`?

yaugenst-flex reviewed (commented):
Nice, looks great! Only some minor comments.

lucas-flexcompute commented on tests/test_data/test_monitor_data.py:
```diff
@@ -865,3 +867,225 @@ def test_no_nans():
+
+    def read_beam_file(self, filename: str) -> tuple:
+        """Read in a Zemax Beam file
+        Adapted from https://community.zemax.com/code-exchange-10/python-class-to-open-zbf-file-4845
```
Should we worry about the (lack of) license in the original code? Is it assumed to be public domain because it was shared in a forum? I'm not sure if that's how it works.

bzhangflex replied:
Good question, I'm not sure either 😅

bzhangflex replied:
I reworked the read zbf function and put it here https://github.com/flexcompute/tidy3d/blob/88a7a20fb11e9a5f6cd84dd04815e8b1feccb747/tidy3d/components/data/zbf.py#L8

lucas-flexcompute commented on tests/test_data/test_monitor_data.py:
```diff
@@ -865,3 +867,225 @@ def test_no_nans():
+        # beamfilename = self.open_zbf()
+        f = open(filename, "rb")
+        # zemax version number
+        version = unpack("i", f.read(4))[0]
```
This use of `unpack` is not [setting alignment and size](https://docs.python.org/3/library/struct.html#byte-order-size-and-alignment), so it will only work if the native settings match the ZBF specs. In my original implementation I fixed the sizes and left native order (`=`) because the manual did not specify it, but it should probably be set little-endian (`<`).

Also, it could probably be a lot more efficient if it unpacked more values at once:
```python
version, nx, ny, ispol, units = unpack("<5i", f.read(4 * 5))
```

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+        mode_index : int
+            Mode index selection.
+        num_samples: int
+            Number of samples to use. Must be a power of 2
```
could you explain a bit more what is being sampled? and what does the default value correspond to?

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+        e_x = e_x.sel(f=freq, drop=True)
+        e_y = e_y.sel(f=freq, drop=True)
+
+        if "mode_index" in e_x.coords:
```
what case does this cover? I guess if the object is a `ModeSolverData`?  

If the `mode_index` is only used in this case, maybe it does actually make sense for the default value to be `None`? And then only do the `isel` here if it's not `None`?

@yaugenst-flex

bzhangflex replied:
Good question, I assumed it was for `ModeSolverData` but I haven't tested it yet. I'll test it out and also change the default value

tylerflex replied:
if so maybe we can mention this in the changelog too

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+
+        # Use the central frequency if freq is not specified
+        if freq is None:
+            freq = e_x.coords["f"].values[e_x.coords["f"].size // 2]
```
do we need to worry if the number of frequency coords is even? wonder if we should rather interpolate to the mean frequency?

lucas-flexcompute replied:
This is a limitation in Zemax. It must be a power of 2.

bzhangflex replied:
now it interpolates to the mean frequency https://github.com/flexcompute/tidy3d/blob/fe6faac2f3d5b27c573866db970d2e5edd28fb83/tidy3d/components/data/monitor_data.py#L1159

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+        # Header info
+        version = 1
+        polarized = 1
+        units = 0  # "mm": 0, "cm": 1, "in": 2, "m": 3
```
should this be something that the user can configure through the kwargs?

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+        version = 1
+        polarized = 1
+        units = 0  # "mm": 0, "cm": 1, "in": 2, "m": 3
+        lda = C_0 / freq * 1e-3
```
this 1e-3 floating around here and below makes me a bit nervous. I think it would be better if we had something like 

```py
# kwarg length_unit = "mm"

length_scale_map = dict(mm=1e-3, cm=1e-2, in=...) # maps kwarg to length scale factor

if length_unit not in length_scale_map:
    raise KeyError(...)

length_scale = length_scale_map[length_unit]

```

and then use `length_scale` in place of 1e-3.

If there are free units floating around like this it can lead to some unexpected bugs.

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+        r_y = 0
+
+        # Pilot beam z position w.r.t. the waist
+        z_x = 0 * 1e-3
```
what's with the 0 * 1e-3? are we planning to allow this to be modified eventually?

bzhangflex replied:
good catch! I'm not fully sure what the pilot beam is used for (seems like it helps [guide the actual beam](https://community.zemax.com/got-a-question-7/pilot-and-beam-width-549)) but it sounds like a good idea to let the user define values for it.

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+        # Number of samples: n_x = 2 ** a, a >= 5 and a <= 13 (32-bit), or a <= 14 (64-bit)
+        if num_samples is not None:
+            # custom user defined number of samples
+            # Raise an error if the number of samples is not a power of 2
```
can you explain why the num_samples needs to be a factor of 2? in principle we could interpolate using however points they provide?

bzhangflex replied:
it's a requirement set by zemax. My guess is it might have to do with FFTs in their beam propagation algorithm.

Should we not enforce it here and let the user choose an arbitrary number? That could be fine too if Zemax emits a clear error message, or changes behavior in the future

tylerflex replied:
it's probably fine to enforce it here. maybe we can just mention in the error message that this is a requirement for zemax

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1224,6 +1225,127 @@ def make_adjoint_sources(
+            n_x = num_samples
+            n_y = num_samples
+        else:
+            n_x = 2 ** min(13, max(5, int(np.log2(x.size) + 1)))
```
should these `5` and `13` be hardcoded somewhere? eg as a module level constant? or configurable? not sure exactly where they come from. is it a requirement of zemax?

bzhangflex replied:
hmm it was left in from @lucas-flexcompute's code, I wonder if there was a specific reason for the values. If not maybe we don't specify any default `n_x` or `n_y`?

lucas-flexcompute replied:
The Zemax manual I have (dated July 8, 2011) defines that the number of samples must be a power of 2 between 32 and 8192 (inclusive)

tylerflex commented on CHANGELOG.md:
it might be worth also adding this to the FAQ section of the docs, as we recently had some users unaware of our export to MATLAB. so probably some extra information about this capability in the docs wouldn't hurt. Also, not sure if it applies, but it might be nice also to add it to at least one notebook example.

bzhangflex replied:
maybe dumb question, how do I update the FAQ?

tylerflex replied:
I think you add one [here](https://github.com/flexcompute/tidy3d-faq/tree/9b8479c00a7b2d5fb81a90dd21c7b8e92423ef58/_faqs) and then link it in the proper .rst [file](https://github.com/flexcompute/tidy3d-faq/tree/9b8479c00a7b2d5fb81a90dd21c7b8e92423ef58/docs)

bzhangflex replied:
Created a [PR](https://github.com/flexcompute/tidy3d-faq/pull/29), keeping as a draft until this is merged and the docs are updated, so I can link to the source in the FAQ

tylerflex reviewed (changes requested):
Thanks @bzhangflex , this looks great! Will be super useful and a highly requested feature!

I had a few minor comments. I'm not familiar with zemax configuration so take what I say with a bit of a grain of salt.

tomflexcompute commented:
Since this is a bit experimental and largely untested, there might be cases that it will not work properly. Should we add a warning when it's called? Something like "This is an experimental feature. If you run into any issues, contact us at flexcompute.com/tidy3d/technical-support/"

bzhangflex commented:
Pushed quite a few changes
- Created a read_zbf() utility function https://github.com/flexcompute/tidy3d/blob/fe6faac2f3d5b27c573866db970d2e5edd28fb83/tidy3d/components/data/zbf.py#L8
- Added `FieldDataset.from_zbf()` for creating a new `FieldDataset` from a zbf file
- In `to_zbf()`, require the user to choose the # of field samples `n_x` and `n_y` separately. Checks that they are powers of two between 2^5 and 2^13 per Zemax manual
- If `freq` in `to_zbf()` is not specified, default to the mean of the frequencies
- `units` is now a kwarg
- User has the option to specify pilot beam parameters. Defaults are 0
- Updated changelog
- Updated warning messages to include support url
- use emulated_run in the pytests

tylerflex commented on tidy3d/components/data/dataset.py:
```diff
@@ -257,6 +258,90 @@ class FieldDataset(ElectromagneticFieldDataset):
+            by zbf data.
+        """
+        log.warning(
+            "`FieldDataset.from_zbf()` is currently an experimental feature."
```
it's confusing but for console outputs you will want the code to be in regular single quotes like '

tylerflex replied:
this comes up in the ValueErrors and things below, so just double check all of them 👍

tylerflex replied:
https://www.notion.so/flexcompute/Python-Style-Guide-afac5dbbf949463e89c18fe39b7453c9

bzhangflex replied:
Thanks! I think I've corrected all the formatting issues

tylerflex commented on tidy3d/components/data/dataset.py:
```diff
@@ -257,6 +258,90 @@ class FieldDataset(ElectromagneticFieldDataset):
+        """
+        log.warning(
+            "`FieldDataset.from_zbf()` is currently an experimental feature."
+            " If any issues are encountered, please contact Flexcompute support https://www.flexcompute.com/tidy3d/technical-support/"
```
let's also put the url in the single quotes

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1066,6 +1068,157 @@ def translated_copy(self, vector: Coordinate) -> ElectromagneticFieldData:
+        background_refractive_index: float,
+        n_x: int,
+        n_y: int,
+        units: UnitsZBF,
```
do we want to set defaults for units or background ref_index? (not sure just wondering)

lucas-flexcompute replied:
Just a minor comment: in the previous implementation, they did have default values, which I think is great to reduce user overload. Same for `n_x` and `n_y`. However, they were provided based on "reasonable defaults", which can be improved depending on the use case.

bzhangflex replied:
good point, added the defaults back in, they are probably good to have

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1066,6 +1068,157 @@ def translated_copy(self, vector: Coordinate) -> ElectromagneticFieldData:
+            Number of field samples along x. Must be a power of 2, between 2^5 and 2^13 inclusive per Zemax's requirements.
+        n_y : int
+            Number of field samples along y. Must be a power of 2, between 2^5 and 2^13 inclusive per Zemax's requirements.
+        units : `UnitsZBF`
```
double check the style guide. I think we use double backslash quotes inside of the docstrings or anything rendered by docs basically.  Also would be good to add to `.zbf` where it comes up.

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1066,6 +1068,157 @@ def translated_copy(self, vector: Coordinate) -> ElectromagneticFieldData:
+        x = e_x.coords[dim1].values
+        y = e_x.coords[dim2].values
+
+        # Use the mean frequency if freq is not specified
```
would it be worth warning in this case? or not needed?

bzhangflex replied:
good idea, I added a warning

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1066,6 +1068,157 @@ def translated_copy(self, vector: Coordinate) -> ElectromagneticFieldData:
+        # If the data is ModeData, choose one of the modes to save
+        if "mode_index" in e_x.coords:
+            if mode_index is None:
+                raise ValueError("`mode_index` is required for `ModeData.to_zbf()`")
```
regular single quotes

tylerflex commented on tidy3d/components/data/zbf.py:
```diff
@@ -0,0 +1,128 @@
+import numpy as np
+
+
+def read_zbf(filename: str):
```
add return type annotation, just `dict` is probably fine

tylerflex commented on tidy3d/components/data/zbf.py:
```diff
@@ -0,0 +1,128 @@
+    Ex = Ex_real + 1j * Ex_imag
+    Ey = Ey_real + 1j * Ey_imag
+
+    return {
```
this is probably not a "must do" but more a "highly recommended": To make things more robust, I'd create a simple data structure for it in pydantic, 

```py
class ZBFData(Tidy3DBaseModel):
    version : str = pd.Field(..., title='Version', description=...)
    ...
```
and then you can return an instance of this instead of a `dict`

basically it offloads what wrote out by hand docstring to a model that we can reference in the doc and will do some validation and things. In general it's a nice approach we use to keep it more "tidy" when passing data with a lot of fields around. otherwise it can get quite complex

bzhangflex replied:
good call, I created a pydantic data structure, let me know if it looks ok https://github.com/flexcompute/tidy3d/blob/7533e425c22dfcd540a9d4514c5ca1cc3a2f7cd8/tidy3d/components/data/zbf.py#L13

tylerflex reviewed (changes requested):
Looking like it's almost there! mostly minor comments

tylerflex commented on tidy3d/constants.py:
```diff
@@ -238,6 +238,7 @@
         "mm": 1e-3,
         "cm": 1e-4,
         "m": 1e-6,
+        "in": 1.0 / 25400,
```
this data structure is a bit unfamiliar to me. I was confused about what it was mapping. I guess for a key `k`, the value is basically how much 1 um is in units of `k`? maybe can you add a short sentence explaining a bit more?

bzhangflex replied:
yup, updated the docstring

bzhangflex commented:
Thanks @tylerflex!, let me know how it looks now

tylerflex reviewed (approved):
looks good, thanks @bzhangflex !

tylerflex commented on tidy3d/components/data/dataset.py:
```diff
@@ -257,6 +258,92 @@ class FieldDataset(ElectromagneticFieldDataset):
+        See Also
+        --------
+        :class:`.ZBFData':
+            A class containing data read in from a `.zbf` file.
```
double backslash here.

bzhangflex replied:
oops thanks

tylerflex commented on tidy3d/components/data/dataset.py:
```diff
@@ -257,6 +258,92 @@ class FieldDataset(ElectromagneticFieldDataset):
     )
 
+    def from_zbf(filename: str, dim1: xyz, dim2: xyz) -> FieldDataset:
+        """Creates a :class:`.FieldDataset` from a Zemax Beam File (`.zbf`).
```
double backslash here.
